Make fatal DiagnosticBuilder yield `!`

compiler-errors opened the pull request:
Fatal errors should really be fatal, so emitting them should cause us to exit at the same time. 

Fine with just throwing away these changes if they're not worthwhile. Also, maybe we want to use an uninhabited enum instead of `!`. 

r? @eddyb who has been working on `DiagnosticBuilder` stuff, feel free to reassign.

compiler-errors commented on compiler/rustc_typeck/src/structured_errors/missing_cast_for_variadic_arg.rs:
```diff
@@ -21,7 +21,7 @@ impl<'tcx> StructuredDiagnostic<'tcx> for MissingCastForVariadicArg<'tcx> {
 
     fn diagnostic_common(&self) -> DiagnosticBuilder<'tcx, ErrorGuaranteed> {
-        let mut err = self.sess.struct_span_fatal_with_code(
+        let mut err = self.sess.struct_span_err_with_code(
```
I don't think this was intended to be fatal.

compiler-errors commented on compiler/rustc_typeck/src/astconv/mod.rs:
```diff
@@ -688,7 +688,7 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
         self.ast_path_to_mono_trait_ref(
             trait_ref.path.span,
-            trait_ref.trait_def_id().unwrap_or_else(|| FatalError.raise()),
+            trait_ref.trait_def_id().unwrap(),
```
I don't see any value in `unwrap_or_else(|| FatalError.raise())`, though I think this could be an `.expect` call if we want a better message.

compiler-errors replied:
nvm, this is to suppress ICEs I guess

compiler-errors commented on compiler/rustc_errors/src/lib.rs:
```diff
@@ -7,6 +7,7 @@
 #![feature(backtrace)]
 #![feature(if_let_guard)]
 #![feature(let_else)]
+#![feature(never_type)]
```
I don't know if it's worthwhile introducing more cases of `never_type` since T-lang has discussed cutting back some of the specialness of the type.

compiler-errors reviewed (commented):
Also curious if we should prevent `Fatal` level errors from being created with a regular `DiagnosticBuilder<ErrorGuaranteed>`. Because in that case, they really should just be `Error` level errors.

hellow554 commented on compiler/rustc_errors/src/diagnostic_builder.rs:
```diff
@@ -198,6 +198,45 @@ impl EmissionGuarantee for () {
+    /// Creates a new `DiagnosticBuilder` with an already constructed
+    /// diagnostic.
+    crate fn new_diagnostic_fatal(handler: &'a Handler, diagnostic: Diagnostic) -> Self {
+        debug!("Created new diagnostic");
```
is there a reason for this debug?

compiler-errors replied:
lol I just copied it from the impl for ErrorGuaranteed above

hellow554 commented on compiler/rustc_errors/src/diagnostic_builder.rs:
```diff
@@ -198,6 +198,45 @@ impl EmissionGuarantee for () {
+impl<'a> DiagnosticBuilder<'a, !> {
+    /// Convenience function for internal use, clients should use one of the
+    /// `struct_*` methods on [`Handler`].
+    crate fn new_fatal(handler: &'a Handler, level: Level, message: &str) -> Self {
```
`new_fatal` with level begin `Level::Fatal` feels redundant to me. What's the reason for providing the `level` argument instead of hardcoding it to `Level::Fatal`?

compiler-errors replied:
copypasta is the reason, I can fix

hellow554 commented on compiler/rustc_errors/src/lib.rs:
```diff
@@ -764,15 +765,15 @@ impl Handler {
-    pub fn struct_fatal(&self, msg: &str) -> DiagnosticBuilder<'_, ErrorGuaranteed> {
-        DiagnosticBuilder::new_guaranteeing_error::<{ Level::Fatal }>(self, msg)
+    pub fn struct_fatal(&self, msg: &str) -> DiagnosticBuilder<'_, !> {
+        DiagnosticBuilder::new_fatal(self, Level::Fatal, msg)
```
regarding `level`, see above

bors commented:
:umbrella: The latest upstream changes (presumably #95056) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

hellow554 commented:
hey @compiler-errors do you had time to resolve my concerns?

oli-obk commented:
Please consider how such statements put pressure on contributors. Everyone is welcome to contribute at their own pace. The triage team has its own set of guidelines for how frequently to remind contributors that there is outstanding work to be done. Personally, I believe we should leave such reminders to the triage team.

eddyb commented:
@bors r+

bors commented:
:pushpin: Commit d1babd779127abdc3fa878c90e6de488dd3b8414 has been approved by `eddyb`

<!-- @bors r=eddyb d1babd779127abdc3fa878c90e6de488dd3b8414 -->
<!-- homu: {"type":"Approved","sha":"d1babd779127abdc3fa878c90e6de488dd3b8414","approver":"eddyb"} -->

bors commented:
:hourglass: Testing commit d1babd779127abdc3fa878c90e6de488dd3b8414 with merge 3df6b768eccf7b891a9e7f3abc974dd4b9712481...
<!-- homu: {"type":"BuildStarted","head_sha":"d1babd779127abdc3fa878c90e6de488dd3b8414","merge_sha":"3df6b768eccf7b891a9e7f3abc974dd4b9712481"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/5714902821?check_suite_focus=true)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/5714902821?check_suite_focus=true","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-debug`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/5714868846?check_suite_focus=true) [(plain)](https://github.com/rust-lang-ci/rust/commit/3df6b768eccf7b891a9e7f3abc974dd4b9712481/checks/5714868846/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_errors v0.0.0 (/checkout/compiler/rustc_errors)
error[E0308]: mismatched types
   --> compiler/rustc_errors/src/diagnostic_builder.rs:230:41
    |
230 |                 handler.emit_diagnostic(&db.inner.diagnostic);
    |                                         ^^^^^^^^^^^^^^^^^^^^ types differ in mutability
    = note: expected mutable reference `&mut diagnostic::Diagnostic`
    = note: expected mutable reference `&mut diagnostic::Diagnostic`
                       found reference `&Box<diagnostic::Diagnostic>`
For more information about this error, try `rustc --explain E0308`.
error: could not compile `rustc_errors` due to previous error
warning: build failed, waiting for other jobs to finish...
error: build failed
```

</details>

compiler-errors commented:
Rebased and fixed error due to #95149 (changed a `&` to a `&mut` [here](https://github.com/rust-lang/rust/pull/94789/files#diff-6c98b29d710fd5dbb3073d32b31b83021fdd67d5eb7713027c4b3b421423e623R230))

oli-obk commented:
@bors r=eddyb

bors commented:
:pushpin: Commit 928388bad20572e9ffa575319bb6d3b636bcdc69 has been approved by `eddyb`

<!-- @bors r=eddyb 928388bad20572e9ffa575319bb6d3b636bcdc69 -->
<!-- homu: {"type":"Approved","sha":"928388bad20572e9ffa575319bb6d3b636bcdc69","approver":"eddyb"} -->

bors commented:
:hourglass: Testing commit 928388bad20572e9ffa575319bb6d3b636bcdc69 with merge 0e4524e5b4a9c5656fef69f532f96eb9959a1803...
<!-- homu: {"type":"BuildStarted","head_sha":"928388bad20572e9ffa575319bb6d3b636bcdc69","merge_sha":"0e4524e5b4a9c5656fef69f532f96eb9959a1803"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/5721277902?check_suite_focus=true)
Approved by: eddyb
Pushing 0e4524e5b4a9c5656fef69f532f96eb9959a1803 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"eddyb","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/5721277902?check_suite_focus=true"},"merge_sha":"0e4524e5b4a9c5656fef69f532f96eb9959a1803"} -->

rust-timer commented:
Finished benchmarking commit (0e4524e5b4a9c5656fef69f532f96eb9959a1803): [comparison url](https://perf.rust-lang.org/compare.html?start=b3e46a9763d86a1681ddaaf0e60ce537302b9c92&end=0e4524e5b4a9c5656fef69f532f96eb9959a1803).

**Summary**: This benchmark run did not return any relevant results. 10 results were found to be statistically significant but too small to be relevant.

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



@rustbot label: -perf-regression
